Add -smtpAuth for mail servers that require authentication

logan opened the pull request:
I'm trying out alertmanager in Amazon EC2, and the easiest delivery of notifications I could think of is through Amazon's email sending service. It requires auth, though, so I threw this change together.

brian-brazil commented:
That'd put a password on the command line, you should read the password from a file.

I use a local postfix in null client mode http://www.postfix.org/STANDARD_CONFIGURATION_README.html#null_client

discordianfish commented:
I would prefer and ENV variable. It's easier to configure and at similar secure.

logan commented:
How does SMTP_AUTH sound?

logan commented:
Or maybe I should go for a more complete implementation.

If SMTP_AUTH_USERNAME is given, then check for the AUTH and TLS extensions and a compatible auth mechanism supported by the server.

If the server supports PLAIN auth and SMTP_AUTH_PASSWORD is given, then use PLAIN auth. An optional SMTP_AUTH_IDENTITY will also be passed to the server.

If the server supports CRAM-MD5 auth and SMTP_AUTH_SECRET is given, then use CRAM-MD5 auth.

If neither of these conditions is met, then plain old unencrypted, unauthenticated mail sending takes place.

juliusv commented:
That sounds good to me (although I'm not an email auth expert, so I'll trust you there or wait for other comments). Also, :+1: to env vars for credentials.

logan commented:
Ok, this change is ready for another round of review. I've updated the title of the pull request to reflect this new approach.

I don't have a good way to test CRAM-MD5 or the SMTP_AUTH_IDENTITY variable, but I can confirm that PLAIN auth works for me with SMTP_AUTH_USERNAME and SMTP_AUTH_PASSWORD set.

juliusv commented on manager/notifier.go:
```diff
@@ -184,6 +188,42 @@ func (n *notifier) sendEmailNotification(to string, a *Alert) error {
+				if err := c.Auth(smtp.CRAMMD5Auth(username, secret)); err != nil {
+					return fmt.Errorf("cram-md5 auth failed: %s", err)
+				}
+				break
```
Wouldn't this `break` only leave the `switch`? From the Go spec: 'A "break" statement terminates execution of the innermost "for", "switch", or "select" statement within the same function.'

I guess your intention is to break out of the `for` loop (not that it would matter in most cases), which could be done via a label?

juliusv commented on manager/notifier.go:
```diff
@@ -184,6 +188,42 @@ func (n *notifier) sendEmailNotification(to string, a *Alert) error {
+				identity := os.Getenv("SMTP_AUTH_IDENTITY")
+
+				// PLAIN auth requires TLS to be started first.
+				host, _, _ := net.SplitHostPort(*smtpSmartHost)
```
To be 100% safe, could you also check the error return value here?

juliusv commented:
:+1: besides comments.

logan commented:
Good catch on the switch break issue, I keep forgetting that Go shares that with C. Instead of a label I've added a default continue case, and moved the break to after the switch. If you'd prefer the labeled breaks I can switch to that easily enough.

discordianfish commented:
LGTM :+1:

juliusv commented:
:+1: Thanks!
